onnxruntime/core/providers/shared_library/provider_bridge_provider.cc

```diff
@@ -0,0 +1,287 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// This is the provider DLL side of the provider API to let providers be built as a DLL
+
+#include "provider_author.h"
+#include <assert.h>
+#include <mutex>
+#include <iostream>  // For std::cout used in a stub
+
+onnxruntime::ProviderHost* g_host{};
+
+#define PROVIDER_NOT_IMPLEMENTED ORT_THROW("Unimplemented shared library provider method");
+
+namespace onnxruntime {
+
+void SetProviderHost(ProviderHost& host) {
+  g_host = &host;
+}
+
+struct OnUnloadFunction {
+  OnUnloadFunction(std::function<void()> function) : function_(std::move(function)) {}
+
+  std::function<void()> function_;
+  bool enabled_{true};
+};
+
+static std::unique_ptr<std::vector<std::unique_ptr<OnUnloadFunction>>> s_run_on_unload_;
+
```

Comment:
Does the vector need to be managed by a unique_ptr?

Reply:
I guess it doesn't need to be. I was getting used to doing that for the thread_locals as they don't get destroyed on dll unload. Trying it without it...

Reply:
Now I remember why. The allocator hasn't been initialized yet. So this must not allocate memory until the host has been set.